rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,279 @@
+}
+
+struct OriginatingSource {
+    receiver: tokio::sync::mpsc::Receiver<EventFinalization>,
+    identifier: String,
+}
+
+struct EventSource {
+    source: Arc<OriginatingSource>,
+    id: EventId,
+}
+
+struct EventMetadata {
+    // existing fields
+    sources: Box<\[EventSource\]>,
+}
+```
+
+## Doc-level Proposal
+
+The following additional source configuration will be added:
+
+```toml
+# Global enable option
+acknowledgements = true
+
+[sources.my_source_id]
+  # Per-source enable option
+  acknowledgements = true
+```
+
+We would also need to document when acknowledgement happens for each source.
+
+## Rationale
+
+The above structure provides for several considerations:
+
+1.  This the minimum amount of data that can be added to the metadata to
+    fully support this feature.
+
+2.  Each source element is fixed size and requires no additional
+    allocations.
+
+3.  If a source does not need or is not configured to require
+    finalization, it will not contribute to the list of sources and so
+    has no additional event overhead, and no additional allocations when
+    the event is created.
+
+4.  Sending the finalization status to the source does not require any
+    lookups or topology traversal.
+
+5.  No additional work is required to handle dropped sources due to
+    topology reconfiguration, other than the expected checking for a
+    closed channel when sending.
+
+## Drawbacks
+
+1.  This adds a substantial base size overhead to each event (minimum of
+    two words), even for configurations that do not support or require
+    end-to-end acknowledgement.
+
+## Alternatives
+
+1.  The set of sources could be stored in a more customary
+    `Vec<EventSource>`. This provides for merging multiple sources with
+    a minimum of code. However, the data added to the event structure
+    then grows by an additional word.
+
+2.  The originating source could be stored as simply the unique
+    identifier string. This requires that all reporting of delivery
+    status proceed through a dictionary lookup instead of simply sending
+    it through a channel, increasing the run-time overhead.
+
+3.  Due to the variety of source event identifiers, two other
+    possibilities for the `SourceId` type would be either a `Box<dyn
+    SourceIdTrait>` or a `Box<str>`. Both of these are only two words
+    long, which will likely be smaller than the final size of `enum
+    SourceId` once all required variants are covered. However, since
+    both of these are variably sized, they would require an additional
+    allocation, negating any benefits they would provide in flexibility.
+    Additionally, while a string is a simple and well understood data
+    type, storing the required data in a string requires additional
+    serializing and deserializing steps to use the data.
+
+## Outstanding Questions
+
+1. Is this feature intended to be opt-in to maximize performance in the
+   default configuration, or opt-out to maximize reliability?
```

Comment:
I'm a little concerned about this not being consistently defaulted across configurations. For example, people are already confused about the back-pressure behavior of one source -> sink vs. one source -> n sinks. I think we should default it either on or off for all topologies.

Reply:
This was the purpose of the global `acknowledgements` option I briefly mentioned above. I will clarify that.

Reply:
I definitely agree it's a concern, I'm just not sure how best to address it without potentially adding to the confusion. We can say that it's "on" for all configurations, but (1) there are many configurations where it doesn't apply, and (2) "on" by default puts us in exactly the same situation that backpressure is in right now, where users wouldn't get acks if one out of n sinks is slow.

The comparison to backpressure is a good one, because these are essentially the same problem. If we can come up with a default behavior that makes it clear what effect each sink has on the flow of data, I'd be in favor of implementing it.

That being said, maybe it's not a problem? We can document that it's on by default but that doesn't mean it actually has to do anything for cases where it isn't applicable, and behavior consistent with backpressure is at the very least consistent.